### PR TITLE
Some update.

### PR DESCRIPTION
**Pull request by miskcoo**
* Move variance_normalizing to variance_reduction in reinforce,
* Locally cache unhashable object in VariationalObjective.

### CODE REVIEW

File: zhusuan/legacy/variational_legacy.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ def nvil(log_joint,
          latent,
          baseline=None,
          decay=0.8,
-         variance_normalization=False,
+         variance_reduction=True,
```
Comment (thjashin) [inline zhusuan/legacy/variational_legacy.py]: It's better to follow the order <variance_reduction, baseline, decay>, and the logic would be if not variance_reduction, do nothing. If variance_reduction, by default use moving mean, and under this, if baseline, use baseline.

File: zhusuan/legacy/variational_legacy.py [hunk 2]
```diff
@@ -151,7 +151,7 @@ def nvil(log_joint,
         by `log_joint`. A trainable estimation for the scale of the
         variational lower bound, which is typically dependent on observed
         values, e.g., a neural network with observed values as inputs.
-    :param variance_normalization: Whether to use variance normalization.
+    :param variance_reduction: Whether to use variance reduction.
```
Comment (thjashin) [inline zhusuan/legacy/variational_legacy.py]: More explanation: By default will subtract the learning signal with a moving mean estimation of it. Users can pass an additional customized baseline using the `baseline` argument, in that way the returned will be a tuple of costs, the former for the gradient estimator, the latter for adapting the baseline.

And for baseline param: This will be additional

File: zhusuan/legacy/variational_legacy.py [hunk 3]
```diff
@@ -177,21 +177,14 @@ def nvil(log_joint,
     if variance_normalization is True:
```
Comment (thjashin) [inline zhusuan/legacy/variational_legacy.py]: bug?

File: zhusuan/variational/exclusive_kl.py [hunk 4]
```diff
@@ -182,27 +181,18 @@ def reinforce(self,
             l_signal = l_signal - baseline
             cost += baseline_cost
```
Comment (thjashin) [inline zhusuan/variational/exclusive_kl.py]: separate gradient estimator cost with baseline cost, if baseline passed, return a tuple

File: zhusuan/variational/base.py [hunk 5]
```diff
@@ -49,35 +49,50 @@ def __init__(self, log_joint, observed, latent):
...
         self._joint_obs = merge_dicts(observed, self._latent_outputs)
-        self._dict_key = (log_joint,
-                          frozenset(latent_k),
-                          frozenset(map(tuple, latent_v)),
-                          frozenset(six.iteritems(observed)))
+        try:
+            self._dict_key = (log_joint,
+                              frozenset(latent_k),
+                              frozenset(map(tuple, latent_v)),
+                              frozenset(six.iteritems(observed)))
```
Comment (thjashin) [inline zhusuan/variational/base.py:56]: will the same dictionary be frozen into the same?

Comment (miskcoo) [reply zhusuan/variational/base.py:56]: yes

File: zhusuan/legacy/variational_legacy.py (hunk 3 above)
Comment (thjashin) [reply zhusuan/legacy/variational_legacy.py]: Sorry I mean the legacy should stick to the old logic and not changed to keep it backward compatible for at least a version.

File: zhusuan/variational/exclusive_kl.py [hunk 6]
```diff
@@ -161,55 +161,54 @@ def reinforce(self,
...
             self._log_joint_term()
         if self._axis is not None:
             cost = tf.reduce_mean(cost, self._axis)
-        return cost
+            if baseline is not None:
+                baseline_cost = tf.reduce_mean(baseline_cost, self._axis)
+
+        if baseline is not None:
+            return cost, baseline_cost
+        else: return cost
```
Comment (thjashin) [inline zhusuan/variational/exclusive_kl.py]: new line
